src/components/loading_strategy.ts

```diff
@@ -36,93 +31,63 @@ function defaultOnPlatformApiError(error: Error) {
-  const req = genClientProxyRequest({ proxy, uiMessageWebviewUrlScheme, language, widgetType, options })
-
-  useEffect(() => {
-    makeClientProxyRequest(buildProxyRequest(req))
-      .then((json) => setWidgetUrl(json.widget_url.url))
-      .catch(onProxyRequestError)
-  }, [])
-
-  return widgetUrl
-}
-
-export function usePlatformApiSso<Options>({
-  apiKey,
-  clientId,
-  userGuid,
-  environment,
-  widgetType,
-  options,
-  uiMessageWebviewUrlScheme,
-  language,
-  onSsoError = defaultOnPlatformApiError,
-}: LoadingParams<PlatformApiLoadingProps, Options>) {
+export function useWidgetUrl<Props extends WidgetLoadingProps & WidgetOptionProps, Opts>(
+  widgetType: Type,
+  props: Props,
+  optsFromProps: (ps: Props) => Opts,
+): string | null {
   const [widgetUrl, setWidgetUrl] = useState<string | null>(null)
 
-  const params = buildPlatformApiRequestParams<Options>(apiKey, clientId, userGuid,
-    environment, widgetType, uiMessageWebviewUrlScheme, language, options)
-
   useEffect(() => {
-    makePlatformApiRequest(params)
-      .then((response) => setWidgetUrl(response.widget_url.url))
-      .catch((error) => onSsoError(error))
+    if (isLoadingWithUrl(props)) {
+      setWidgetUrl(props.url)
+    } else if (isLoadingWithClientProxy(props)) {
+      const { proxy, language } = props
+      const buildProxyRequest = props.buildProxyRequest || ((req) => req)
+      const onProxyRequestError = props.onProxyRequestError || defaultOnClientProxyRequestError
+      const uiMessageWebviewUrlScheme = props.uiMessageWebviewUrlScheme || ""
+      const options = optsFromProps(props)
+
+      const req = genClientProxyRequest({ proxy, uiMessageWebviewUrlScheme, language, widgetType, options })
+
+      makeClientProxyRequest(buildProxyRequest(req))
+        .then((json) => setWidgetUrl(json.widget_url.url))
+        .catch(onProxyRequestError)
+    } else if (isLoadingWithPlatformApiSso(props)) {
+      const { apiKey, clientId, userGuid, environment, language } = props
+      const onSsoError = props.onSsoError || defaultOnPlatformApiError
+      const uiMessageWebviewUrlScheme = props.uiMessageWebviewUrlScheme || ""
+      const options = optsFromProps(props)
+
+      const params = buildPlatformApiRequestParams<Opts>(apiKey, clientId, userGuid,
+        environment, widgetType, uiMessageWebviewUrlScheme, language, options)
+
+      makePlatformApiRequest(params)
+        .then((response) => setWidgetUrl(response.widget_url.url))
+        .catch((error) => onSsoError(error))
+    } else {
+      isLoadingWithBadProps()
+    }
   }, [])
 
   return widgetUrl
 }
-
-export function useWidgetUrl<Props extends WidgetLoadingProps & WidgetOptionProps, Opts>(
-  widgetType: Type,
-  props: Props,
-  optsFromProps: (ps: Props) => Opts,
-): string | null {
-  if (isLoadingWithUrl(props)) {
-    return props.url
-  } else if (isLoadingWithClientProxy(props)) {
-    return useClientProxy({
```

Comment:
I _think_ conditionally calling hooks here was actually OK since the condition would never change and we'd always take the same branch, but this change only makes this code more stable so I'm all for it.